pkg/bloomgateway/bloomgateway.go

```diff
@@ -325,17 +352,24 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
 			level.Debug(g.logger).Log("msg", "got partial result", "task", task.ID, "tenant", tenantID, "fp_int", uint64(res.Fp), "fp_hex", res.Fp, "chunks_to_remove", res.Removals.Len(), "progress", fmt.Sprintf("%d/%d", len(responses), requestCount))
 			// wait for all parts of the full response
 			if len(responses) == requestCount {
-				for _, o := range responses {
-					if res.Removals.Len() == 0 {
-						continue
-					}
-					// we must not remove items from req.Refs as long as the worker may iterater over them
-					g.removeNotMatchingChunks(req, o)
-				}
-				return &logproto.FilterChunkRefResponse{ChunkRefs: req.Refs}, nil
+				break outer
 			}
 		}
 	}
+
+	for _, o := range responses {
+		if o.Removals.Len() == 0 {
+			continue
+		}
+		// we must not remove items from req.Refs as long as the worker may iterater over them
```

Comment:
I believe this comment is not valid anymore
```suggestion
```

Reply:
Well, it's still somewhat valid. The point is, that we cannot remove items as soon as we receive a partial result from the channel, but we have to collect them into slice and once we have all results, remove them.

Reply:
some kinda ... yep, we can leave it, just as info for the future.. 
my point was that we moved out this block and we do not run this code when we receive the results from worker, we run this code when we get all the results